### PR TITLE
*: Randomize lease placement in ALTER TABLE SCATTER

### LINKED ISSUE #17341: sql, storage: SCATTER doesn't do anything
```sql
root@:26257/ro> ALTER TABLE devices SCATTER;
+-----+--------+
| key | pretty |
+-----+--------+
+-----+--------+
(0 rows)
root@:26257/ro> SHOW TESTING_RANGES FROM TABLE devices;
+-----------+---------+----------+--------------+
| Start Key | End Key | Replicas | Lease Holder |
+-----------+---------+----------+--------------+
| NULL      | NULL    | {1,2,3}  |            3 |
+-----------+---------+----------+--------------+
```
The scatter command returns instantly and nothing changes (I can repeat the above many times and always see the same lease holder).

```
root@:26257/ro> alter table raw_visits scatter;
+-----+--------+
| key | pretty |
+-----+--------+
+-----+--------+
(0 rows)
root@:26257/ro> SHOW TESTING_RANGES FROM TABLE raw_visits;
+-----------+---------+----------+--------------+
| Start Key | End Key | Replicas | Lease Holder |
+-----------+---------+----------+--------------+
| NULL      | /1      | {1,2,3}  |            2 |
| /1        | /10     | {1,2,3}  |            1 |
| /10       | /20     | {1,2,3}  |            1 |
| /20       | /50     | {1,2,3}  |            1 |
| /50       | /100    | {1,2,3}  |            1 |
| /100      | NULL    | {1,2,3}  |            1 |
+-----------+---------+----------+--------------+
(6 rows)
root@:26257/ro> ALTER TABLE raw_visits SCATTER;
+-----+--------+
| key | pretty |
+-----+--------+
+-----+--------+
(0 rows)
root@:26257/ro> SHOW TESTING_RANGES FROM TABLE raw_visits;
+-----------+---------+----------+--------------+
| Start Key | End Key | Replicas | Lease Holder |
+-----------+---------+----------+--------------+
| NULL      | /1      | {1,2,3}  |            2 |
| /1        | /10     | {1,2,3}  |            1 |
| /10       | /20     | {1,2,3}  |            1 |
| /20       | /50     | {1,2,3}  |            1 |
| /50       | /100    | {1,2,3}  |            1 |
| /100      | NULL    | {1,2,3}  |            1 |
+-----------+---------+----------+--------------+
(6 rows)
```

This is on a 3-node cluster. I expect the lease holders to change when we scatter. Is this no longer the case? @benesch

### PR DESCRIPTION
**Pull request by a-robinson**
This brings us partially back to the v1.0 behavior of the ALTER TABLE
SCATTER command, before it was modified to meet RESTORE's needs.

Randomizing the replicas themselves is much harder to do properly, and
there hasn't been as much demand for it.

Fixes #17341 (assuming we don't consider replica scattering necessary)

Release note: ALTER TABLE SCATTER now randomizes the locations of the
leases for all ranges in the referenced table or index.

### CODE REVIEW

**Comment by RaduBerinde** [comment]
Nice! Thanks for working on this!

:lgtm:

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21431#-:-L2gpOIp65izpCd60I5i:b-lwlety)*

File: pkg/sql/scatter_test.go [hunk 1]
```diff
@@ -47,44 +48,64 @@ func TestScatter(t *testing.T) {
...
+		} else if len(cols) != 5 {
+			return nil, fmt.Errorf("expected 4 columns, got %#v", cols)
+		}
+		vals := []interface{}{
+			new(interface{}),
+			new(interface{}),
+			new(int),
+			new(interface{}),
+			new(int),
+		}
```
Comment (benesch) [inline pkg/sql/scatter_test.go]: nit: this is getting pretty gross. What about this?

```go
rows := r.Query(t, `SELECT "Range ID", "Lease Holder" FROM [SHOW TESTING_RANGES FROM TABLE test.t]`)
for ; rows.Next(); numRows++ {
  var rangeID, leaseholder int
  if err := rows.Scan(&rangeID, &leaseholder); err != nil {
    // ...
  }
  // ...
}
```

**Review by benesch — APPROVED** [review]

**Comment by petermattis** [comment]
:lgtm: (though I only looked at the implementation code, not the test updates)

---

Reviewed 2 of 2 files at r1.
Review status: 1 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica_command.go, line 2100 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21431#-L2goBgfXE-VrRBP9zJ5-r2-2100:-L2zQ1rQqJQWsypgeq18:b3tsiuj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e17e69d0ff46a2d4594b9049a4c6abfec050e3d8/pkg/storage/replica_command.go#L2100)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

This causes two lease transfers—is there a way to adjust the `allowLeaseTransfer` logic above to skip the final requeue when `args.RandomizeLeases` is true?

(Perhaps there's some reason that won't work that I'm missing?)
</blockquote></details>

We'll only enter the block below if the replicate queue didn't transfer the lease. So only 1 lease transfer will occur. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21431#-:-L2zQHgnhSzPuQSW9AlK:bgbl00d)*

**Comment by benesch** [comment]
Review status: 1 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica_command.go, line 2100 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21431#-L2goBgfXE-VrRBP9zJ5-r2-2100:-L2zuvYTIRcBngOUVNS5:blenoah) ([raw file](https://github.com/cockroachdb/cockroach/blob/e17e69d0ff46a2d4594b9049a4c6abfec050e3d8/pkg/storage/replica_command.go#L2100)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

We'll only enter the block below if the replicate queue didn't transfer the lease. So only 1 lease transfer will occur. 
</blockquote></details>

Oh, good point. Ignore me.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21431)*

**Comment by a-robinson** [comment]
Review status: 1 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/scatter_test.go, line 65 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21431#-L2goBgfXE-VrRBP9zJ--r2-65:-L2zvoCwMivk5sEJLefQ:b-dtkdrq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e17e69d0ff46a2d4594b9049a4c6abfec050e3d8/pkg/sql/scatter_test.go#L65)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

nit: this is getting pretty gross. What about this?

```go
rows := r.Query(t, `SELECT "Range ID", "Lease Holder" FROM [SHOW TESTING_RANGES FROM TABLE test.t]`)
for ; rows.Next(); numRows++ {
  var rangeID, leaseholder int
  if err := rows.Scan(&rangeID, &leaseholder); err != nil {
    // ...
  }
  // ...
}
```
</blockquote></details>

Done, thanks.

---

*[pkg/storage/replica_command.go, line 2100 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21431#-L2goBgfXE-VrRBP9zJ5-r2-2100:-L2zuhs1FklMZnJ47ylf:bpef86) ([raw file](https://github.com/cockroachdb/cockroach/blob/e17e69d0ff46a2d4594b9049a4c6abfec050e3d8/pkg/storage/replica_command.go#L2100)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

We'll only enter the block below if the replicate queue didn't transfer the lease. So only 1 lease transfer will occur. 
</blockquote></details>

Correct - the `if` below verifies that the local replica still owns the lease, so it'll only run if we didn't transfer the lease above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21431)*

**Comment by benesch** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r1, 4 of 5 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21431#-:-L2zxUCUJo0wHO6NU3sh:bnfp4nl)*

**Pull request merged**